disks: Redo "Add disk" and "Insert media" dialog with new framework

mvollmer opened the pull request:
This untangles the complex logic in the "Add disk" dialog. It keeps the three modes cleanly separate and avoids overlapping their states and logic. For example, the "format" handling for the "Custom path" mode is different from that of "Use existing", but it was hard to see which code is for which.

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-a14a96d9-20251211-132844-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-a3e5ddd4-20251211-152938-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
+        } else
+            return [];
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+    forEach(func: (val: DialogValue<ArrayElement<T>>, index: number) => void): void {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            val.forEach((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.#dialog._rename_validation_state(this.#path, j + 1, j);
+            this.set(toSpliced(val, index, 1) as T);
```
These 6 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
+                                        } else {
+                                            console.log("OUTDATED", this.#path);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+
+    _get_validation_state(path: string): DialogValidationState | null {
+        if (path in this.#validation_state)
+            return this.#validation_state[path];
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1343 @@
+        if (path in this.#validation_state)
+            return this.#validation_state[path];
+        else
+            return null;
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251215-171634-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

mvollmer commented:
The height change in the dialog is a bit sus: https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251216-120651-fedora-43/pixeldiff.html#TestMachinesDisks-testAddDiskAdditionalOptions-vm-add-disk-modal-additional-options

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+function toSpliced<T>(arr: T[], start: number, deleteCount: number, ...rest: T[]): T[] {
+    const copy = [...arr];
+    copy.splice(start, deleteCount, ...rest);
+    return copy;
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+    map<X>(func: (val: DialogValue<ArrayElement<T>>, index: number) => X): X[] {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+    forEach(func: (val: DialogValue<ArrayElement<T>>, index: number) => void): void {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            val.forEach((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+        if (Array.isArray(val)) {
+            for (let j = index; j < val.length - 1; j++)
+                this.#dialog._rename_validation_state(this.#path, j + 1, j);
+            this.set(toSpliced(val, index, 1) as T);
```
These 6 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+    add(item: ArrayElement<T>) {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            this.set(val.concat(item) as T);
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+                                            console.log("DONE VALIDATE ASYNC", this.#path, result);
+                                            this.#dialog._set_validation_result(this.#path, val, result);
+                                        } else {
+                                            console.log("OUTDATED", this.#path);
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+
+    _get_validation_state(path: string): DialogValidationState | null {
+        if (path in this.#validation_state)
+            return this.#validation_state[path];
```
These 3 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+        if (path in this.#validation_state)
+            return this.#validation_state[path];
+        else
+            return null;
```
This added line is not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1382 @@
+            const state = this.#validation_state[path];
+            if (state.timeout_id) {
+                console.log("CANCEL TIMEOUT", path);
+                window.clearTimeout(state.timeout_id);
```
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-e2dfc49b-20251217-161440-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-12f2a464-20251217-173453-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+    map<X>(func: (val: DialogValue<ArrayElement<T>>, index: number) => X): X[] {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            return val.map((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-7ce1b174-20251217-183650-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+    forEach(func: (val: DialogValue<ArrayElement<T>>, index: number) => void): void {
+        const val = this.get();
+        if (Array.isArray(val)) {
+            val.forEach((_, i) => func(this.sub(i as keyof T) as DialogValue<ArrayElement<T>>, i));
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-7ce1b174-20251217-183650-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-7ce1b174-20251217-183650-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+            state.round_id = this.#get_current_validation_round_id();
+            debug("cache hit", state.path, state.cached_result);
+            this.#set_validation(state.path, state.cached_result);
+            return true;
```
These 5 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-7ce1b174-20251217-183650-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous commented on src/components/common/dialog.tsx:
```diff
@@ -0,0 +1,1550 @@
+        if (state.timeout_id) {
+            debug("timeout cancel", state.path);
+            window.clearTimeout(state.timeout_id);
+            state.timeout_id = 0;
```
These 4 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-7ce1b174-20251217-183650-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-806f6cea-20251222-084707-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

cockpituous reviewed (commented):
There are more than 10 code coverage comments, see [the full report here](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-bae37c6c-20251222-091539-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html).

tomasmatus reviewed (changes requested):
Phew that was long, it's pretty hard to look at the diff so I also looked at the code in my editor "as-is" and I think it makes sense. I take the minor changes to tests that no features were lost in this dialog rework. Bellow are some things that I noticed but I think I'll need to go over this one more time.

mvollmer commented on src/components/vm/disks/diskAdd.tsx:
```diff
@@ -74,907 +77,667 @@ function getDiskUsageMessage(vms: VM[], storagePool: StoragePool, volumeName: st
-        initiallySelected = "empty";
-    }
+    if (!bus_types.includes(bus_type))
+        bus_types.push(bus_type);
```
Hmm, no, this is just the initialization value.  The similar `update_AdditionalOptions` function below does use `set()`.

mvollmer replied:
But, the weird `_fileInfoCache` construction for the `CustomPath` component _is_ modified without `set()`, but I would argue that this is ok since it is not part of the visual state of the dialog. But it's maybe too subtle.

I had the idea that a validation function can return an arbitrary result in addition to field errors. I think that might turn out to be useful in a number of cases. I haven't come up with a great API for that, though...

mvollmer commented on src/components/vm/disks/diskAdd.tsx:
```diff
@@ -74,907 +77,667 @@ function getDiskUsageMessage(vms: VM[], storagePool: StoragePool, volumeName: st
-                fieldId={`${idPrefix}-file-autocomplete`}
-                label={_("Custom path")}>
+                label={_("Custom path")}
+                id={field.sub("file").id()}
```
No, it's more complicated.  We need a proper "id" on the FormGroup because Browser.set_file_autocomplete` needs to be given the FormGroup, not the actual widget. We could also have a separate field(Id/id combo on the FormGroup and the widget, but since it doesn't work anyway, I had removed it. Can we keep it that way? I have added a comment to explain this a bit.

mvollmer replied:
Alternatively, we could make a proper wrapper for FileAutoComplete... I want to do that anyway for the "Create VM" dialog.  Let me have a go at that.

mvollmer replied:
Ok, I did that and also factored out a SizeInput widget. We are gonna use that in more places as well.

mvollmer commented:
> The height change in the dialog is a bit sus: https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2449-2c41faae-20251216-120651-fedora-43/pixeldiff.html#TestMachinesDisks-testAddDiskAdditionalOptions-vm-add-disk-modal-additional-options

This is caused by the fact that now there is only a `HelperText` thing when there is actually something to show, and not always.  I think the new behavior is more correct.

tomasmatus commented on src/components/storagePools/storageVolumeCreateBody.tsx:
```diff
@@ -6,179 +6,124 @@
+        const poolCapacity = convertToUnit(_storagePool.capacity, units.B, v.unit);
+        if (!v.size || isNaN(Number(v.size)))
+            return _("Size must be a number");
+        console.log("SZ", v.size, Number(v.size));
```
let's drop the debug print :)

mvollmer replied:
upps!

tomasmatus commented on src/components/vm/disks/diskAdd.tsx:
```diff
@@ -74,907 +77,661 @@ function getDiskUsageMessage(vms: VM[], storagePool: StoragePool, volumeName: st
-    };
+            if (values.mode === CREATE_NEW && typeof values.create_new != "string") {
+                const params = values.create_new;
+                const size = Math.round(convertToUnit(params.volume.newSize.size, params.volume.newSize.unit, 'MiB'));
```
Not sure about this `Math.round` you force pushed. Does it "discard" the decimal size I set? Or is it some libvirt limitation?

I create disk with 1.5GiB, UI shows it as 1GiB
I create disk with 1.51GiB, UI shows it as 1.5GiB
... but these are rounding errors elsewhere in the "Disks" card

mvollmer replied:
It's a limitation in the libvirt API. It wants integer megabytes.  But a 1.5 GiB input leading to 1 GiB display of disk size looks wrong. You should maybe get 1.4 GiB but not something as far off as 1 GiB.

mvollmer replied:
But you are right, of course. When I create a new volume with an input of "1.5" and unit "GiB", it will have a capacity of 1610612736 bytes. This is exactly 1536 MiB.

The "Disks" table of a VM will show that as "Size 1 GiB". The "Storage volumes" tab of the "Storage pools" page will show it as "Size 1.5 GB".  (The "Size" of a pool is given in "GiB", not "GB".)

This seems messed up. Let's look at that separately.

mvollmer replied:
I have moved the `Math.round` call to `createVolumeXML` so that not all callers have to know about this restriction.

tomasmatus replied:
> This seems messed up. Let's look at that separately.

Agree, let's do that.

> I have moved the Math.round call to createVolumeXML so that not all callers have to know about this restriction.

Sounds good!

tomasmatus reviewed (changes requested):
Don't see a reason to block this so I won't make things up. Please fix the merge conflict and see the two comments bellow.

mvollmer committed:
disks: Reimplement "Add Disk" and "Create Volume" dialogs

Using the new dialog framework. This also includes the "Insert media"
variant of the "Add disk" dialog.

- Instead of showing an empty and disabled dropdown select control,
  the radio button that would have shown it is now disabled.

- The size input now allows decimal numbers and doesn't spontaneously
  erase the input when a non-numeric character is typed.

tomasmatus reviewed (approved):
Looks good, thanks!
